{"cells":[{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":["import os\n","import re\n","import pandas as pd\n","from tqdm import tqdm\n"]},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["                 tissue           variant_id            gene_id\n","0  Adipose Subcutaneous   chr1_64764_C_T_b38  ENSG00000227232.5\n","1  Adipose Subcutaneous  chr1_665098_G_A_b38  ENSG00000227232.5\n","2  Adipose Subcutaneous  chr1_666028_G_A_b38  ENSG00000227232.5\n","3  Adipose Subcutaneous  chr1_108826_G_C_b38  ENSG00000269981.1\n","4  Adipose Subcutaneous  chr1_126108_G_A_b38  ENSG00000269981.1\n","(71478479, 3)\n"]}],"source":["# the user should have downloaded the GTEx v8 eQTL data\n","# eqtl_dir: the directory contains the 'XXX.v8.signif_variant_gene_pairs.txt.gz' files\n","eqtl_dir = r\"/path/to/your/GTEx/v8/eQTL\"\n","df_all = []\n","for filename in os.listdir(eqtl_dir):\n","    if \".signif_variant_gene_pairs.\" not in filename:\n","        continue\n","    tissue = filename.split(\".\", maxsplit=1)[0].replace(\"_\", \" \")\n","    filename = os.path.join(eqtl_dir, filename)\n","    if not os.path.isfile(filename):\n","        continue\n","    df_eqtl = pd.read_csv(\n","        filename,\n","        delimiter=\"\\t\",\n","        compression=\"gzip\",\n","        usecols=[\"variant_id\", \"gene_id\"],\n","    )\n","    df_eqtl.drop_duplicates(inplace=True)\n","    df_eqtl.insert(0, \"tissue\", tissue)\n","    df_all.append(df_eqtl)\n","df_all = pd.concat(df_all, axis=0, sort=False, ignore_index=True)\n","print(df_all.head())\n","print(df_all.shape)\n"]},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["  chromosome  type  start    end strand            gene_id\n","0       chr1  gene  11869  14403      +  ENSG00000223972.5\n","2       chr1  exon  11869  12227      +  ENSG00000223972.5\n","3       chr1  exon  12613  12721      +  ENSG00000223972.5\n","4       chr1  exon  12975  13052      +  ENSG00000223972.5\n","5       chr1  exon  13221  14403      +  ENSG00000223972.5\n","(384871, 6)\n"]}],"source":["# read the GTF given by GTEx\n","# gtf_path: the path of 'gencode.v26.GRCh38.genes.gtf' file given by GTEx\n","gtf_path = \"/path/to/your/GTEx/v8/gencode.v26.GRCh38.genes.gtf\"\n","df_gtf = pd.read_csv(\n","    gtf_path, delimiter=\"\\t\", comment=\"#\", header=None, usecols=[0, 2, 3, 4, 6, 8]\n",")\n","# extract the gene IDs\n","df_gtf[\"gene_id\"] = [re.search(r'gene_id \"(.+?)\"', i).group(1) for i in df_gtf[8]]\n","del df_gtf[8]\n","df_gtf.columns = [\"chromosome\", \"type\", \"start\", \"end\", \"strand\", \"gene_id\"]\n","df_gtf = df_gtf[df_gtf[\"type\"].isin([\"exon\", \"gene\"])]\n","print(df_gtf.head())\n","print(df_gtf.shape)\n"]},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["100%|██████████| 34548/34548 [1:41:50<00:00,  5.65it/s]   \n"]},{"name":"stdout","output_type":"stream","text":["                 tissue              variant_id             gene_id  \\\n","0  Adipose Subcutaneous  chrX_100548181_A_G_b38  ENSG00000000003.14   \n","1  Adipose Subcutaneous  chrX_100573255_C_T_b38  ENSG00000000003.14   \n","2  Adipose Subcutaneous  chrX_100592861_G_A_b38  ENSG00000000003.14   \n","3  Adipose Subcutaneous  chrX_100594020_G_A_b38  ENSG00000000003.14   \n","4  Adipose Subcutaneous  chrX_100594054_T_C_b38  ENSG00000000003.14   \n","\n","  variant_position  variant_distance  \n","0       downstream             78928  \n","1       downstream             53854  \n","2       downstream             34248  \n","3       downstream             33089  \n","4       downstream             33055  \n","(71478479, 5)\n"]}],"source":["# calculate the distances of eQTLs to eGenes\n","df_all[[\"chromosome\", \"position\"]] = [\n","    [(j := i.split(\"_\", maxsplit=2))[0], int(j[1])] for i in df_all[\"variant_id\"]\n","]\n","df_dist = []\n","for gene_id, df_temp in tqdm(df_all.groupby(\"gene_id\")):\n","    # get the positions of gene and its exons, respectively\n","    gene_pos = df_gtf[df_gtf[\"gene_id\"] == gene_id].copy().drop(columns=[\"gene_id\"])\n","    exon_pos = gene_pos[gene_pos[\"type\"] == \"exon\"]\n","    gene_pos = gene_pos[gene_pos[\"type\"] == \"gene\"]\n","    gene_pos = gene_pos.iloc[0, :].to_dict()\n","    variant_positions = []\n","    variant_distances = []\n","    for _, row in df_temp.iterrows():\n","        if gene_pos[\"strand\"] == \"+\":\n","            tss_distance = row[\"position\"] - gene_pos[\"start\"]\n","            if row[\"position\"] < gene_pos[\"start\"]:\n","                # upstream\n","                variant_positions.append(\"upstream\")\n","                variant_distances.append(gene_pos[\"start\"] - row[\"position\"])\n","            elif row[\"position\"] > gene_pos[\"end\"]:\n","                # downstream\n","                variant_positions.append(\"downstream\")\n","                variant_distances.append(row[\"position\"] - gene_pos[\"end\"])\n","            else:\n","                if (\n","                    len(\n","                        exon_pos[\n","                            (exon_pos[\"start\"] <= row[\"position\"])\n","                            & (exon_pos[\"end\"] >= row[\"position\"])\n","                        ]\n","                    )\n","                    > 0\n","                ):\n","                    # exon\n","                    variant_positions.append(\"exon\")\n","                else:\n","                    # intron\n","                    variant_positions.append(\"intron\")\n","                variant_distances.append(row[\"position\"] - gene_pos[\"start\"])\n","        else:\n","            tss_distance = row[\"position\"] - gene_pos[\"end\"]\n","            if row[\"position\"] > gene_pos[\"end\"]:\n","                # upstream\n","                variant_positions.append(\"upstream\")\n","                variant_distances.append(row[\"position\"] - gene_pos[\"end\"])\n","            elif row[\"position\"] < gene_pos[\"start\"]:\n","                # downstream\n","                variant_positions.append(\"downstream\")\n","                variant_distances.append(gene_pos[\"start\"] - row[\"position\"])\n","            else:\n","                if (\n","                    len(\n","                        exon_pos[\n","                            (exon_pos[\"start\"] <= row[\"position\"])\n","                            & (exon_pos[\"end\"] >= row[\"position\"])\n","                        ]\n","                    )\n","                    > 0\n","                ):\n","                    # exon\n","                    variant_positions.append(\"exon\")\n","                else:\n","                    # intron\n","                    variant_positions.append(\"intron\")\n","                variant_distances.append(gene_pos[\"end\"] - row[\"position\"])\n","    df_temp[\"variant_position\"] = variant_positions\n","    df_temp[\"variant_distance\"] = variant_distances\n","    df_temp = df_temp.drop(columns=[\"chromosome\", \"position\"])\n","    df_dist.append(df_temp.copy())\n","df_dist = pd.concat(df_dist, axis=0, sort=False, ignore_index=True)\n","print(df_dist.head())\n","print(df_dist.shape)\n","df_dist.to_csv(\n","    \"eQTL_distance.txt.gz\",\n","    sep=\"\\t\",\n","    compression=\"gzip\",\n","    index=False,\n",")\n","del df_dist\n"]},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["(71478479, 4)\n","               variant_id             gene_id variant_position  \\\n","0  chrX_100548181_A_G_b38  ENSG00000000003.14       downstream   \n","1  chrX_100573255_C_T_b38  ENSG00000000003.14       downstream   \n","2  chrX_100592861_G_A_b38  ENSG00000000003.14       downstream   \n","3  chrX_100594020_G_A_b38  ENSG00000000003.14       downstream   \n","4  chrX_100594054_T_C_b38  ENSG00000000003.14       downstream   \n","\n","   variant_distance  \n","0             78928  \n","1             53854  \n","2             34248  \n","3             33089  \n","4             33055  \n","(13791909, 4)\n"]}],"source":["# load the table calculated above and drop duplicate eQTL-eGene pairs\n","df = pd.read_csv(\n","    \"eQTL_distance.txt.gz\",\n","    delimiter=\"\\t\",\n","    compression=\"gzip\",\n","    usecols=[\"gene_id\", \"variant_id\", \"variant_position\", \"variant_distance\"],\n",")\n","print(df.shape)\n","df.drop_duplicates(subset=[\"gene_id\", \"variant_id\"], inplace=True)\n","print(df.head())\n","print(df.shape)\n"]},{"cell_type":"code","execution_count":8,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["              gene_id  gene_length\n","0   ENSG00000223972.5         2535\n","6   ENSG00000227232.5        15144\n","39  ENSG00000186092.4          918\n","42  ENSG00000238009.6        39929\n","52  ENSG00000233750.3         3812\n","(34548, 2)\n"]}],"source":["# calculate the lengths of genes\n","genes = set(df[\"gene_id\"])\n","df_gene_len = df_gtf[df_gtf[\"type\"] == \"gene\"][[\"gene_id\", \"start\", \"end\"]].copy()\n","df_gene_len = df_gene_len[df_gene_len[\"gene_id\"].isin(genes)]\n","df_gene_len[\"gene_length\"] = df_gene_len[\"end\"] - df_gene_len[\"start\"] + 1\n","df_gene_len.drop(columns=[\"start\", \"end\"], inplace=True)\n","print(df_gene_len.head())\n","print(df_gene_len.shape)\n"]},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["100%|██████████| 34548/34548 [00:16<00:00, 2144.94it/s]"]},{"name":"stdout","output_type":"stream","text":["              gene_id  exon_length\n","0  ENSG00000000003.14         4535\n","1   ENSG00000000005.5         1610\n","2  ENSG00000000419.12         1207\n","3  ENSG00000000457.13         5586\n","4  ENSG00000000460.16         4579\n","(34548, 2)\n"]},{"name":"stderr","output_type":"stream","text":["\n"]}],"source":["# calculate the lengths of exons\n","df_exon_len = df_gtf[df_gtf[\"type\"] == \"exon\"][[\"gene_id\", \"start\", \"end\"]].copy()\n","df_exon_len = df_exon_len[df_exon_len[\"gene_id\"].isin(genes)]\n","gene_ids = []\n","lengths = []\n","for gene_id, df_temp in tqdm(df_exon_len.groupby(\"gene_id\")):\n","    pos = set()\n","    for _, row in df_temp.iterrows():\n","        pos.update(range(row[\"start\"], row[\"end\"] + 1))\n","    gene_ids.append(gene_id)\n","    lengths.append(len(pos))\n","df_exon_len = pd.DataFrame({\"gene_id\": gene_ids, \"exon_length\": lengths})\n","print(df_exon_len.head())\n","print(df_exon_len.shape)\n"]},{"cell_type":"code","execution_count":11,"metadata":{},"outputs":[{"data":{"text/plain":["0.0012718772280142147"]},"execution_count":11,"metadata":{},"output_type":"execute_result"}],"source":["# the density of eQTLs on exons\n","# calculate the count of eQTLs on exons\n","df_gene = df[df[\"variant_position\"].isin([\"exon\", \"intron\"])]\n","df_exon = df_gene[df_gene[\"variant_position\"] == \"exon\"]\n","df_nvariants = df_exon[[\"gene_id\", \"variant_id\"]].groupby(\"gene_id\").count()\n","# The gene without any eQTLs on its exons will have a zero count\n","genes_novar = sorted(genes - set(df_nvariants.index))\n","df_nvariants = pd.concat(\n","    [\n","        df_nvariants,\n","        pd.DataFrame({\"variant_id\": [0] * len(genes_novar)}, index=genes_novar),\n","    ],\n","    axis=0,\n","    sort=False,\n","    ignore_index=False,\n",")\n","# density = total count / total length\n","df_nvariants = pd.merge(df_nvariants, df_exon_len, left_index=True, right_on=\"gene_id\")\n","density_exon = df_nvariants[\"variant_id\"].sum() / df_nvariants[\"exon_length\"].sum()\n","density_exon\n"]},{"cell_type":"code","execution_count":12,"metadata":{},"outputs":[{"data":{"text/plain":["0.0011245422861891033"]},"execution_count":12,"metadata":{},"output_type":"execute_result"}],"source":["# the density of eQTLs on introns\n","# calculate the count of eQTLs on introns\n","df_intron = df_gene[df_gene[\"variant_position\"] == \"intron\"]\n","df_nvariants = df_intron[[\"gene_id\", \"variant_id\"]].groupby(\"gene_id\").count()\n","df_nvariants.shape\n","# The gene without any eQTLs ont its introns will have a zero count\n","genes_novar = sorted(genes - set(df_nvariants.index))\n","df_nvariants = pd.concat(\n","    [\n","        df_nvariants,\n","        pd.DataFrame({\"variant_id\": [0] * len(genes_novar)}, index=genes_novar),\n","    ],\n","    axis=0,\n","    sort=False,\n","    ignore_index=False,\n",")\n","df_nvariants.shape\n","# density = total count / total length\n","df_nvariants = pd.merge(df_nvariants, df_gene_len, left_index=True, right_on=\"gene_id\")\n","df_nvariants = pd.merge(df_nvariants, df_exon_len, on=\"gene_id\")\n","df_nvariants[\"intron_length\"] = (\n","    df_nvariants[\"gene_length\"] - df_nvariants[\"exon_length\"]\n",")\n","density_intron = df_nvariants[\"variant_id\"].sum() / df_nvariants[\"intron_length\"].sum()\n","density_intron\n"]},{"cell_type":"code","execution_count":13,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["121925.0 121388.0\n","0.0007142395971031626\n"]}],"source":["# density of eQTLs in median distance up/downstream\n","# formula: The number of all up/downstream eQTLs * 0.5 / ((Dup + Ddown) * number of genes)\n","df_nogene = df[df[\"variant_position\"].isin([\"upstream\", \"downstream\"])]\n","# The number of up/downstream eQTLs\n","nvariants = len(df_nogene)\n","# median distances up/downstream\n","dist_up = df_nogene[df_nogene[\"variant_position\"] == \"upstream\"][\n","    \"variant_distance\"\n","].median()\n","dist_down = df_nogene[df_nogene[\"variant_position\"] == \"downstream\"][\n","    \"variant_distance\"\n","].median()\n","print(dist_up, dist_down)\n","# number of genes\n","ngenes = len(genes)\n","density_up_down_stream = nvariants * 0.5 / (dist_up + dist_down) / ngenes\n","print(density_up_down_stream)\n"]},{"cell_type":"code","execution_count":14,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["1.780743091215802\n","1.5744608542428344\n"]}],"source":["# ratio of densities\n","print(density_exon / density_up_down_stream)\n","print(density_intron / density_up_down_stream)\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Python 3.9.1 ('env4base')","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.1"},"orig_nbformat":4,"vscode":{"interpreter":{"hash":"6f9d0e333f0135229d628c3554584ce24b7e667acf50311d63438641208afc17"}}},"nbformat":4,"nbformat_minor":2}
