{"cells":[{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["import os\n","import re\n","import pandas as pd\n","from multiprocessing import Pool\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# read gene positions\n","# The GTF file can be downloaded from GENCODE (https://www.gencodegenes.org/human/)\n","# version 36, GRCh37 reference genome\n","gtf_path = \"/path/to/your/gencode.v36lift37.annotation.gtf.gz\"\n","df_gtf = pd.read_csv(\n","    gtf_path,\n","    delimiter=\"\\t\",\n","    comment=\"#\",\n","    header=None,\n","    usecols=[0, 2, 3, 4, 6, 8],\n","    compression=\"gzip\",\n",")\n","# extract gene IDs\n","gene_id_pattern = re.compile(r'gene_id \"(ENSG[0-9]+)\\.')\n","df_gtf[\"gene_id\"] = [gene_id_pattern.search(i).group(1) for i in df_gtf[8]]\n","del df_gtf[8]\n","df_gtf.columns = [\n","    \"chromosome\",\n","    \"type\",\n","    \"start\",\n","    \"end\",\n","    \"strand\",\n","    \"gene_id\",\n","]\n","# only use chromosome 1-22 and X\n","df_gtf = df_gtf[\n","    df_gtf[\"chromosome\"].isin([f\"chr{i}\" for i in list(range(1, 23)) + [\"X\"]])\n","]\n","\n","df_gene_pos = df_gtf[df_gtf[\"type\"] == \"gene\"].copy()\n","df_exon_pos = df_gtf[df_gtf[\"type\"] == \"exon\"].copy()\n","del df_gtf\n","del df_gene_pos[\"type\"]\n","del df_exon_pos[\"type\"]\n","print(df_gene_pos.head())\n","print(df_gene_pos.shape)\n","print(df_exon_pos.head())\n","print(df_exon_pos.shape)\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# convert the gene position information to dict like \"{gene ID->{gene position}}\"\n","gene_pos_dict = {\n","    gid: gene_pos.iloc[0, :-1].to_dict()\n","    for gid, gene_pos in df_gene_pos.groupby(\"gene_id\")\n","}\n","del df_gene_pos\n","# convert exon position information to dict like \"{gene ID->exon position dataframe}\"\n","exon_pos_dict = {\n","    gid: exon_pos.drop(columns=[\"gene_id\"])\n","    for gid, exon_pos in df_exon_pos.groupby(\"gene_id\")\n","}\n","del df_exon_pos\n","# for example\n","print(gene_pos_dict[\"ENSG00000223972\"])\n","print(exon_pos_dict[\"ENSG00000223972\"])\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# [In] 8\n","# {gene ID->gene length}\n","gene_len = {gid: v[\"end\"] - v[\"start\"] + 1 for gid, v in gene_pos_dict.items()}\n","# {gene ID->exon length}\n","exon_len = {}\n","for gid, df_exon in exon_pos_dict.items():\n","    pos = set()\n","    for start, end in zip(df_exon[\"start\"], df_exon[\"end\"]):\n","        pos.update(range(start, end + 1))\n","    exon_len[gid] = len(pos)\n","# {gene ID->intron length}\n","intron_len = {gid: l - exon_len[gid] for gid, l in gene_len.items()}\n","# for example\n","print(gene_len[\"ENSG00000000003\"])\n","print(exon_len[\"ENSG00000000003\"])\n","print(intron_len[\"ENSG00000000003\"])\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["def _get_variant_type_and_distance(\n","    gene_id: str, var_chrom: str, var_pos: int\n",") -> tuple[str, int]:\n","    \"\"\"\n","    giving a variant and a gene,\n","    get the variant type (exon, intron, or up/downstream) and distance\n","\n","    Args:\n","        gene_id (str): ensembl ID.\n","        var_chrom (str): the chromosome of variant.\n","        var_pos (int): the position of variant.\n","\n","    Returns:\n","        tuple[str, int]: variant type (exon, intron, or up/downstream) and distance\n","    \"\"\"\n","    try:\n","        gene_pos = gene_pos_dict[gene_id]\n","        exon_pos = exon_pos_dict[gene_id]\n","        assert var_chrom == gene_pos[\"chromosome\"]\n","        var_type = \"\"\n","        var_dist = -1\n","        if gene_pos[\"strand\"] == \"+\":\n","            if var_pos < gene_pos[\"start\"]:\n","                # upstream\n","                var_type = \"upstream\"\n","                var_dist = gene_pos[\"start\"] - var_pos\n","            elif var_pos > gene_pos[\"end\"]:\n","                # downstream\n","                var_type = \"downstream\"\n","                var_dist = var_pos - gene_pos[\"end\"]\n","            else:\n","                if any(\n","                    (\n","                        start <= var_pos <= end\n","                        for start, end in zip(exon_pos[\"start\"], exon_pos[\"end\"])\n","                    )\n","                ):\n","                    # exon\n","                    var_type = \"exon\"\n","                else:\n","                    # intron\n","                    var_type = \"intron\"\n","                var_dist = var_pos - gene_pos[\"start\"]\n","        else:\n","            if var_pos > gene_pos[\"end\"]:\n","                # upstream\n","                var_type = \"upstream\"\n","                var_dist = var_pos - gene_pos[\"end\"]\n","            elif var_pos < gene_pos[\"start\"]:\n","                # downstream\n","                var_type = \"downstream\"\n","                var_dist = gene_pos[\"start\"] - var_pos\n","            else:\n","                if any(\n","                    (\n","                        start <= var_pos <= end\n","                        for start, end in zip(exon_pos[\"start\"], exon_pos[\"end\"])\n","                    )\n","                ):\n","                    # exon\n","                    var_type = \"exon\"\n","                else:\n","                    # intron\n","                    var_type = \"intron\"\n","                var_dist = gene_pos[\"end\"] - var_pos\n","        return var_type, var_dist\n","    except:\n","        return \"\", -1\n","\n","\n","def get_variant_type_and_distance(args) -> tuple[str, int]:\n","    return _get_variant_type_and_distance(*args)\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["if not os.path.isdir(output_dir := os.path.join(\"eQTL_distance\")):\n","    os.mkdir(output_dir)\n","# path to the significant eQTLs of TCGA\n","# which can be downloaded in this repo\n","eqtl_dir = \"/path/to/your/signif_cis_eQTL\"\n","for eqtl_file in os.listdir(eqtl_dir):\n","    cancer_type = eqtl_file.split(\".\", maxsplit=1)[0]\n","    if os.path.isfile(output_file := os.path.join(output_dir, f\"{cancer_type}.txt.gz\")):\n","        continue\n","    df_eqtl = pd.read_csv(\n","        os.path.join(eqtl_dir, eqtl_file), delimiter=\"\\t\", compression=\"gzip\"\n","    )\n","    df_eqtl[[\"chrom\", \"pos\"]] = [\n","        [(j := i.split(\"_\", maxsplit=2))[0], int(j[1])] for i in df_eqtl[\"variant_id\"]\n","    ]\n","    df_dist = df_eqtl[[\"variant_id\", \"gene_id\"]].copy()\n","    with Pool(4) as p:\n","        rst = list(\n","            p.map(\n","                get_variant_type_and_distance,\n","                zip(df_eqtl[\"gene_id\"], df_eqtl[\"chrom\"], df_eqtl[\"pos\"]),\n","            ),\n","            total=len(df_eqtl),\n","        )\n","    df_dist[[\"variant_position\", \"variant_distance\"]] = rst\n","    df_dist.to_csv(\n","        output_file, sep=\"\\t\", compression=\"gzip\", index=False, line_terminator=\"\\n\"\n","    )\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# calculate parameters\n","df_param = []\n","for file in list(os.listdir(output_dir)):\n","    cancer_type = file.split(\".\", maxsplit=1)[0].removeprefix(\"TCGA-\")\n","    df = pd.read_csv(\n","        os.path.join(output_dir, file),\n","        delimiter=\"\\t\",\n","        compression=\"gzip\",\n","    )\n","    genes = set(df[\"gene_id\"])\n","\n","    # exon\n","    df_exon = df[df[\"variant_position\"] == \"exon\"]\n","    df_nvariants = df_exon[[\"gene_id\", \"variant_id\"]].groupby(\"gene_id\").count()\n","    genes_novar = sorted(genes - set(df_nvariants.index))\n","    df_nvariants = pd.concat(\n","        [\n","            df_nvariants,\n","            pd.DataFrame({\"variant_id\": [0] * len(genes_novar)}, index=genes_novar),\n","        ],\n","        axis=0,\n","        sort=False,\n","        ignore_index=False,\n","    )\n","    # density = total count / total length\n","    df_nvariants[\"exon_length\"] = df_nvariants.index.map(exon_len)\n","    density_exon = df_nvariants[\"variant_id\"].sum() / df_nvariants[\"exon_length\"].sum()\n","\n","    # intron\n","    df_intron = df[df[\"variant_position\"] == \"intron\"]\n","    df_nvariants = df_intron[[\"gene_id\", \"variant_id\"]].groupby(\"gene_id\").count()\n","    genes_novar = sorted(genes - set(df_nvariants.index))\n","    df_nvariants = pd.concat(\n","        [\n","            df_nvariants,\n","            pd.DataFrame({\"rsID\": [0] * len(genes_novar)}, index=genes_novar),\n","        ],\n","        axis=0,\n","        sort=False,\n","        ignore_index=False,\n","    )\n","    # density = total count / total length\n","    df_nvariants[\"intron_length\"] = df_nvariants.index.map(intron_len)\n","    density_intron = (\n","        df_nvariants[\"variant_id\"].sum() / df_nvariants[\"intron_length\"].sum()\n","    )\n","\n","    # up/downstream\n","    # formula: The number of all up/downstream eQTLs * 0.5 / ((Dup + Ddown) * number of genes)\n","    df_nogene = df[df[\"variant_position\"].isin([\"upstream\", \"downstream\"])]\n","    # The number of up/downstream eQTLs\n","    nvariants = len(df_nogene)\n","    # median distances up/downstream\n","    dist_up = (\n","        df_nogene[df_nogene[\"variant_position\"] == \"upstream\"][\"variant_distance\"]\n","        .median()\n","        .__ceil__()\n","    )\n","    dist_down = (\n","        df_nogene[df_nogene[\"variant_position\"] == \"downstream\"][\"variant_distance\"]\n","        .median()\n","        .__ceil__()\n","    )\n","    # number of genes\n","    ngenes = len(genes)\n","    density_up_down_stream = nvariants * 0.5 / (dist_up + dist_down) / ngenes\n","\n","    # scale the weight of up/downstream into 1\n","    # exon : up/downstream\n","    exon_weight = density_exon / density_up_down_stream\n","    # intron : up/downstream\n","    intron_weight = density_intron / density_up_down_stream\n","    df_param.append([cancer_type, exon_weight, intron_weight, 1.0, dist_up, dist_down])\n","df_param = pd.DataFrame(\n","    df_param,\n","    columns=[\n","        \"cancer_type\",\n","        \"exon_weight\",\n","        \"intron_weight\",\n","        \"up_downstream_weight\",\n","        \"upstream_distance\",\n","        \"downstream_distance\",\n","    ],\n",")\n","df_param\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"env4base","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.1"},"orig_nbformat":4,"vscode":{"interpreter":{"hash":"6f9d0e333f0135229d628c3554584ce24b7e667acf50311d63438641208afc17"}}},"nbformat":4,"nbformat_minor":2}
