{"cells":[{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[],"source":["import os\n","import re\n","import pandas as pd\n","from tqdm import tqdm\n"]},{"cell_type":"code","execution_count":2,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["         rsID      Gene\n","0   rs6004673    CRYBB2\n","1   rs1995969   B3GALTL\n","2   rs2157990    ZSWIM7\n","3    rs202370    ZNF334\n","4  rs10838184  HSD17B12\n","(6432, 2)\n"]}],"source":["# read the eQTLs of ER-positive and ER-negative breast cancers\n","# the supplementary table 2 and table 3 of PMID: 23374354\n","table_files = [\"/path/to/your/supp_table2.xlsx\", \"/path/to/your/supp_table3.xlsx\"]\n","df_eqtl = pd.concat(\n","    [pd.read_excel(file, usecols=[\"rsID\", \"Gene\"]) for file in table_files],\n","    axis=0,\n","    sort=False,\n","    ignore_index=True,\n",")\n","df_eqtl.drop_duplicates(inplace=True)\n","print(df_eqtl.head())\n","print(df_eqtl.shape)\n"]},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["         rsID      Gene chromosome  position\n","0   rs6004673    CRYBB2      chr22  25489386\n","1   rs1995969   B3GALTL      chr13  31300307\n","2   rs2157990    ZSWIM7      chr17  16058679\n","3    rs202370    ZNF334      chr20  46529811\n","4  rs10838184  HSD17B12      chr11  43848310\n","(6421, 4)\n"]}],"source":["# The reference genome used by that paper is NCBI36\n","# We need to get the GRCh38 positions of rsIDs from dbSNP:\n","# https://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/All_20180418.vcf.gz\n","# Here we provide a small file that only contains the involved rsIDs\n","# please download the \"rsID.txt.gz\" from our GitHub repo\n","snp_file = \"/path/to/rsID.txt.gz\"\n","df_snp = pd.read_csv(snp_file, delimiter=\"\\t\", compression=\"gzip\")\n","df_eqtl = pd.merge(df_eqtl, df_snp, on=\"rsID\")\n","print(df_eqtl.head())\n","print(df_eqtl.shape)\n"]},{"cell_type":"code","execution_count":4,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["  chromosome  type  start    end strand gene_name            gene_id\n","0       chr1  gene  11869  14409      +   DDX11L1  ENSG00000223972.5\n","2       chr1  exon  11869  12227      +   DDX11L1  ENSG00000223972.5\n","3       chr1  exon  12613  12721      +   DDX11L1  ENSG00000223972.5\n","4       chr1  exon  13221  14409      +   DDX11L1  ENSG00000223972.5\n","6       chr1  exon  12010  12057      +   DDX11L1  ENSG00000223972.5\n","(1232352, 7)\n"]}],"source":["# read the GTF given by TCGA, which can be downloaded from:\n","# https://api.gdc.cancer.gov/data/25aa497c-e615-4cb7-8751-71f744f9691f\n","# gtf_path: the path of 'gencode.v26.GRCh38.genes.gtf' file given by GTEx\n","gtf_path = \"/path/to/your/gencode.v22.annotation.gtf.gz\"\n","df_gtf = pd.read_csv(\n","    gtf_path,\n","    delimiter=\"\\t\",\n","    comment=\"#\",\n","    header=None,\n","    usecols=[0, 2, 3, 4, 6, 8],\n","    compression=\"gzip\",\n",")\n","# extract gene names\n","df_gtf[\"gene_name\"] = [re.search(r'gene_name \"(.+?)\"', i).group(1) for i in df_gtf[8]]\n","# extract gene IDs\n","df_gtf[\"gene_id\"] = [re.search(r'gene_id \"(.+?)\"', i).group(1) for i in df_gtf[8]]\n","del df_gtf[8]\n","df_gtf.columns = [\n","    \"chromosome\",\n","    \"type\",\n","    \"start\",\n","    \"end\",\n","    \"strand\",\n","    \"gene_name\",\n","    \"gene_id\",\n","]\n","df_gtf = df_gtf[df_gtf[\"type\"].isin([\"exon\", \"gene\"])]\n","# There are some genes with two different IDs\n","# One of them have been currently (Oct. 2022) deprecated or renamed, so keep only the right one\n","# The two ensembl IDs (ENSG00000268500 and ENSG00000105501) of SIGLEC5 are both currently valid\n","# The former is coding gene and the latter is lncRNA, here we retain the former\n","keep_genes = {\n","    \"NOTCH2NL\": \"ENSG000002643423.4\",\n","    \"CFHR2\": \"ENSG00000080910.10\",\n","    \"TREX1\": \"ENSG00000213689.8\",\n","    \"HYMAI\": \"ENSG00000276680.1\",\n","    \"SYT15\": \"ENSG00000204176.12\",\n","    \"RNASE11\": \"ENSG00000173464.13\",\n","    \"MTHFS\": \"ENSG00000136371.8\",\n","    \"CLN3\": \"ENSG00000188603.15\",\n","    \"NDUFA7\": \"ENSG00000267855.4\",\n","    \"KLK9\": \"ENSG00000213022.5\",\n","    \"SIGLEC5\": \"ENSG00000268500.4\",\n","}\n","df_gtf = df_gtf[\n","    (~df_gtf[\"gene_name\"].isin(keep_genes.keys()))\n","    | df_gtf[\"gene_id\"].isin(keep_genes.values())\n","]\n","print(df_gtf.head())\n","print(df_gtf.shape)\n"]},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["100%|██████████| 1491/1491 [01:28<00:00, 16.90it/s]\n"]},{"name":"stdout","output_type":"stream","text":["         rsID    Gene variant_position  variant_distance\n","0  rs16863219   AADAC         upstream            631789\n","1  rs41526349   AADAC         upstream            660638\n","2   rs2870415   AADAC         upstream            607062\n","3  rs12106682   AADAC         upstream            643451\n","4    rs887282  ABCA10       downstream            488447\n","(6000, 4)\n"]}],"source":["# calculate the distances of eQTLs to eGenes\n","df = []\n","for gene_name, df_temp in tqdm(df_eqtl.groupby(\"Gene\")):\n","    gene_pos = (\n","        df_gtf[df_gtf[\"gene_name\"] == gene_name].copy().drop(columns=[\"gene_name\"])\n","    )\n","    if len(gene_pos) == 0:\n","        continue\n","    exon_pos = gene_pos[gene_pos[\"type\"] == \"exon\"]\n","    gene_pos = gene_pos[gene_pos[\"type\"] == \"gene\"]\n","    gene_pos = gene_pos.iloc[0, :].to_dict()\n","    variant_positions = []\n","    variant_distances = []\n","    for _, row in df_temp.iterrows():\n","        if gene_pos[\"strand\"] == \"+\":\n","            tss_distance = row[\"position\"] - gene_pos[\"start\"]\n","            if row[\"position\"] < gene_pos[\"start\"]:\n","                # upstream\n","                variant_positions.append(\"upstream\")\n","                variant_distances.append(gene_pos[\"start\"] - row[\"position\"])\n","            elif row[\"position\"] > gene_pos[\"end\"]:\n","                # downstream\n","                variant_positions.append(\"downstream\")\n","                variant_distances.append(row[\"position\"] - gene_pos[\"end\"])\n","            else:\n","                if (\n","                    len(\n","                        exon_pos[\n","                            (exon_pos[\"start\"] <= row[\"position\"])\n","                            & (exon_pos[\"end\"] >= row[\"position\"])\n","                        ]\n","                    )\n","                    > 0\n","                ):\n","                    # exon\n","                    variant_positions.append(\"exon\")\n","                else:\n","                    # intron\n","                    variant_positions.append(\"intron\")\n","                variant_distances.append(row[\"position\"] - gene_pos[\"start\"])\n","        else:\n","            tss_distance = row[\"position\"] - gene_pos[\"end\"]\n","            if row[\"position\"] > gene_pos[\"end\"]:\n","                # upstream\n","                variant_positions.append(\"upstream\")\n","                variant_distances.append(row[\"position\"] - gene_pos[\"end\"])\n","            elif row[\"position\"] < gene_pos[\"start\"]:\n","                # downstream\n","                variant_positions.append(\"downstream\")\n","                variant_distances.append(gene_pos[\"start\"] - row[\"position\"])\n","            else:\n","                if (\n","                    len(\n","                        exon_pos[\n","                            (exon_pos[\"start\"] <= row[\"position\"])\n","                            & (exon_pos[\"end\"] >= row[\"position\"])\n","                        ]\n","                    )\n","                    > 0\n","                ):\n","                    # exon\n","                    variant_positions.append(\"exon\")\n","                else:\n","                    # intron\n","                    variant_positions.append(\"intron\")\n","                variant_distances.append(gene_pos[\"end\"] - row[\"position\"])\n","    df_temp[\"variant_position\"] = variant_positions\n","    df_temp[\"variant_distance\"] = variant_distances\n","    df_temp = df_temp.drop(columns=[\"chromosome\", \"position\"])\n","    df.append(df_temp.copy())\n","df = pd.concat(df, axis=0, sort=False, ignore_index=True)\n","print(df.head())\n","print(df.shape)\n"]},{"cell_type":"code","execution_count":6,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["      gene_name  gene_length\n","2371     UBE2J2        19977\n","7271       GNB1       105833\n","9211       HES5         1501\n","10099  ARHGEF16        26688\n","18335    APITD1        12367\n","(1380, 2)\n"]}],"source":["# calculate the lengths of genes\n","genes = set(df[\"Gene\"])\n","df_gene_len = df_gtf[df_gtf[\"type\"] == \"gene\"][[\"gene_name\", \"start\", \"end\"]].copy()\n","df_gene_len = df_gene_len[df_gene_len[\"gene_name\"].isin(genes)]\n","df_gene_len[\"gene_length\"] = df_gene_len[\"end\"] - df_gene_len[\"start\"] + 1\n","df_gene_len.drop(columns=[\"start\", \"end\"], inplace=True)\n","print(df_gene_len.head())\n","print(df_gene_len.shape)\n"]},{"cell_type":"code","execution_count":7,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["100%|██████████| 1380/1380 [00:03<00:00, 429.39it/s]"]},{"name":"stdout","output_type":"stream","text":["  gene_name  exon_length\n","0     AADAC         1668\n","1    ABCA10         8405\n","2     ABCC1         7575\n","3    ABCC11         7245\n","4     ABCC4         7819\n","(1380, 2)\n"]},{"name":"stderr","output_type":"stream","text":["\n"]}],"source":["# calculate the lengths of exons\n","df_exon_len = df_gtf[df_gtf[\"type\"] == \"exon\"][[\"gene_name\", \"start\", \"end\"]].copy()\n","df_exon_len = df_exon_len[df_exon_len[\"gene_name\"].isin(genes)]\n","gene_names = []\n","lengths = []\n","for gene_name, df_temp in tqdm(df_exon_len.groupby(\"gene_name\")):\n","    pos = set()\n","    for _, row in df_temp.iterrows():\n","        pos.update(range(row[\"start\"], row[\"end\"] + 1))\n","    gene_names.append(gene_name)\n","    lengths.append(len(pos))\n","df_exon_len = pd.DataFrame({\"gene_name\": gene_names, \"exon_length\": lengths})\n","print(df_exon_len.head())\n","print(df_exon_len.shape)\n"]},{"cell_type":"code","execution_count":8,"metadata":{},"outputs":[{"data":{"text/plain":["2.7843131795464354e-05"]},"execution_count":8,"metadata":{},"output_type":"execute_result"}],"source":["# the density of eQTLs on exons\n","# calculate the count of eQTLs on exons\n","df_gene = df[df[\"variant_position\"].isin([\"exon\", \"intron\"])]\n","df_exon = df_gene[df_gene[\"variant_position\"] == \"exon\"]\n","df_nvariants = df_exon[[\"Gene\", \"rsID\"]].groupby(\"Gene\").count()\n","# The gene without any eQTLs on its exons will have a zero count\n","genes_novar = sorted(genes - set(df_nvariants.index))\n","df_nvariants = pd.concat(\n","    [\n","        df_nvariants,\n","        pd.DataFrame({\"rsID\": [0] * len(genes_novar)}, index=genes_novar),\n","    ],\n","    axis=0,\n","    sort=False,\n","    ignore_index=False,\n",")\n","# density = total count / total length\n","df_nvariants = pd.merge(\n","    df_nvariants, df_exon_len, left_index=True, right_on=\"gene_name\"\n",")\n","density_exon = df_nvariants[\"rsID\"].sum() / df_nvariants[\"exon_length\"].sum()\n","density_exon\n"]},{"cell_type":"code","execution_count":9,"metadata":{},"outputs":[{"data":{"text/plain":["1.0579285196094095e-05"]},"execution_count":9,"metadata":{},"output_type":"execute_result"}],"source":["# the density of eQTLs on introns\n","# calculate the count of eQTLs on introns\n","df_intron = df_gene[df_gene[\"variant_position\"] == \"intron\"]\n","df_nvariants = df_intron[[\"Gene\", \"rsID\"]].groupby(\"Gene\").count()\n","# The gene without any eQTLs ont its introns will have a zero count\n","genes_novar = sorted(genes - set(df_nvariants.index))\n","df_nvariants = pd.concat(\n","    [\n","        df_nvariants,\n","        pd.DataFrame({\"rsID\": [0] * len(genes_novar)}, index=genes_novar),\n","    ],\n","    axis=0,\n","    sort=False,\n","    ignore_index=False,\n",")\n","# density = total count / total length\n","df_nvariants = pd.merge(\n","    df_nvariants, df_gene_len, left_index=True, right_on=\"gene_name\"\n",")\n","df_nvariants = pd.merge(df_nvariants, df_exon_len, on=\"gene_name\")\n","df_nvariants[\"intron_length\"] = (\n","    df_nvariants[\"gene_length\"] - df_nvariants[\"exon_length\"]\n",")\n","density_intron = df_nvariants[\"rsID\"].sum() / df_nvariants[\"intron_length\"].sum()\n","density_intron\n"]},{"cell_type":"code","execution_count":10,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["129110 84394\n","7.889408581830189e-06\n"]}],"source":["# density of eQTLs in median distance up/downstream\n","# formula: The number of all up/downstream eQTLs * 0.5 / ((Dup + Ddown) * number of genes)\n","df_nogene = df[df[\"variant_position\"].isin([\"upstream\", \"downstream\"])]\n","# The number of up/downstream eQTLs\n","nvariants = len(df_nogene)\n","# median distances up/downstream\n","dist_up = (\n","    df_nogene[df_nogene[\"variant_position\"] == \"upstream\"][\"variant_distance\"]\n","    .median()\n","    .__ceil__()\n",")\n","dist_down = (\n","    df_nogene[df_nogene[\"variant_position\"] == \"downstream\"][\"variant_distance\"]\n","    .median()\n","    .__ceil__()\n",")\n","dist_up, dist_down\n","# number of genes\n","ngenes = len(genes)\n","density_up_down_stream = nvariants * 0.5 / (dist_up + dist_down) / ngenes\n","density_up_down_streamdf_nogene = df[\n","    df[\"variant_position\"].isin([\"upstream\", \"downstream\"])\n","]\n","# The number of up/downstream eQTLs\n","nvariants = len(df_nogene)\n","# median distances up/downstream\n","dist_up = (\n","    df_nogene[df_nogene[\"variant_position\"] == \"upstream\"][\"variant_distance\"]\n","    .median()\n","    .__ceil__()\n",")\n","dist_down = (\n","    df_nogene[df_nogene[\"variant_position\"] == \"downstream\"][\"variant_distance\"]\n","    .median()\n","    .__ceil__()\n",")\n","print(dist_up, dist_down)\n","# number of genes\n","ngenes = len(genes)\n","density_up_down_stream = nvariants * 0.5 / (dist_up + dist_down) / ngenes\n","print(density_up_down_stream)\n"]},{"cell_type":"code","execution_count":11,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["3.5291785824844797\n","1.340947814574956\n"]}],"source":["# ratio of densities\n","print(density_exon / density_up_down_stream)\n","print(density_intron / density_up_down_stream)\n"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]}],"metadata":{"kernelspec":{"display_name":"Python 3.9.1 ('env4base')","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.9.1"},"orig_nbformat":4,"vscode":{"interpreter":{"hash":"6f9d0e333f0135229d628c3554584ce24b7e667acf50311d63438641208afc17"}}},"nbformat":4,"nbformat_minor":2}
